Fix issue with layout any and Tstr_eval in the native toplevel

ccasin opened the pull request:
This fixes an issue with using the native toplevel to evaluate expressions with layout `any`, as in:

```
utop # assert false;;
Error: Non-value detected in [value_kind]. 
       Please report this error to the Jane Street compilers team.
       'a has layout any, which is not a sublayout of value.
```

The basic issue is that the native top-level turns `Tstr_eval`s into `Tstr_value`s, and the rest of the compiler expects `Tstr_value`s to be representable.  The fix is just to require `Tstr_eval`s to also be representable.

The observant read will wonder why I added `type_representable_expression` rather than modifying `type_expression`, when no calls to `type_expression` remain in the compiler.  The answer is it's in compilerlibs and used by things like Merlin.

I haven't added a test.  As far as I can tell, there are no tests for the native toplevel, but if someone will point me to them I'll add one.  This does change typing for any `Tstr_eval` in all cases, but to observe it you need to get your hands on something with layout any (which is hard).

goldfirere reviewed (commented):
Turns out you can make a test:

```ocaml
(* TEST
   * toplevel.opt
*)

assert false;;
```

This failed on a nearby build (good). Test it with yours and see what happens. Live discussion suggests you need to repeat the fix.

Otherwise, looks good.

ccasin commented:
OK, I've added two more commits.

The first adds a test.  As we discussed in person, I added this test only for the native code toplevel, because of issues with the testing system that don't seem worth sorting out. (The commit message on that one should say "add native code top-level test", not "... top-level fix", but since it will get squashed away I don't think it's worth fixing)

The second applies the same fix to the `ocaml/toplevel` as was previously applied to `native_toplevel`.  I'm still confused about why we have both `native_toplevel` and `ocaml/toplevel/native`, but at least the test seems to pass on all build configurations.

ccasin commented:
@mshinwell Looks like you need to review this for the change in the native toplevel.

mshinwell reviewed (approved):
Only files owned by me

ccasin commented:
Thanks Mark!  @goldfirere , maybe glance at the changes in the last two commit and then merge this if you're happy.

goldfirere commented:
For the record, I did indeed glance.
